### PR TITLE
Project Report: Numpy Peephole Improvements with JAX

### LINKED ISSUE #331: Project Proposal: Peephole Improvements for numpy using JAX
**What will you do?**
People sometimes encounter numerical stability issues when writing mathematically intensive code, which make it harder for your program to work consistently. There are various tricks that people will use to fix these problems, but if you or your lab-mates don't already know them they can be tricky to find. Of course, many libraries will already implement operations with the tricks, but when you're writing it yourself using basic primitives (for example, if you're trying to implement something from a paper which tells you the mathematical specification but doesn't give you code) it can be tricky to remember/know about all of them.

For example, in neural networks it's pretty common to use `log(sum(exp(x)))` somewhere. However, `sum(exp(x))` can get big very quickly and cause overflow issues, even if `log(sum(exp(x)))` would fit into your float. Luckily, `c + log(sum(exp(x - c))) = log(sum(exp(x)))` for any constant `c`. If you set `c = max(x)`, then would then sum `exp(x-c)` where `x-c <= 0`, avoiding your overflows. See [this blog post](https://www.xarg.org/2016/06/the-log-sum-exp-trick-in-machine-learning/) for a proof.

The goal of this project is to implement a few of these numerical stability tricks as peephole "optimizations" for numpy code. It is the same idea of a peephole optimization in that it looks at the code and finds some small tweaks that it could make to improve it, but the underlying goal is to improve numerical stability, not to replace the code with something equivalent and faster.

**How will you do it?**
[JAX](https://github.com/google/jax) is a library that combines autograd and xla to make it possible to automatically differentiate numpy code and use xla to run it on a GPU, by providing a drop-in replacement for numpy. It provides an intermediate representation called [Jaxpr](https://jax.readthedocs.io/en/latest/jaxpr.html) which lets it represent the mathematical operations in a function, and target it to different platforms. It also provides a tracing-based JIT compiler that lets users decorate a function to be compiled into a faster version.

My plan is to use JAX to create a Jaxpr representation of the code, which I can then search through for opportunities to make peephole improvements.

**How will you empirically measure success?**
This project is going to emphasize correctness more than usefulness, so I will evaluate it by writing test code which checks that the implementation makes its improvements, that it doesn't change code which looks similar but doesn't actually fit into this framework, and that the improvements in fact improve numerical stability.

### PR DESCRIPTION
**Pull request by atucker**
closes #331

Thanks for a great semester! I really enjoyed the class!

### CODE REVIEW

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+# Implementation
+The implementation of the project builds on several ideas explored in class, such as peephole optimizations and intermediate representations.
+
+## Design
+The basic design of the project is to convert code which uses numpy into an intermediate representation, perform the peephole optimizations on the intermediate representation, and then execute the code.
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:17]: I'd probably spell it NumPy, and maybe link to [its homepage](https://numpy.org).

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
++++
+title = "Peephole Improvements for Numpy with JAX"
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:2]: Numpy -> NumPy, probably

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+with is a common way of dealing with this problem, since
+$$\log\sum_{i=1}^n\exp(x_i) = c + \log\sum_{i=1}^n\exp(x_i-c) \text{ for all } c,$$
+and so setting $c = \max_{i\in[n]}(x_i)$ guarantees that you only exponentiate terms
+$$x_i - c \leq 0,$$ avoiding overflow issues. 
+
+#### Logsumexp trick proof
+$$\begin{aligned}\log\sum_{i=1}^n\exp(x_i) &=\log\sum_{i=1}^n1\exp(x_i)\\\ &= \log\sum_{i=1}^n\exp(c-c)\exp(x_i)\\\ &= \log\left(\exp(c)\sum_{i=1}^n\exp(x_i-c)\right)\\\ &= c + \log\sum_{i=1}^n\exp(x_i-c) \end{aligned}$$
+
+### JAX and Jaxpr
+Thankfully, there is already a project which converts numpy code into an intermediate representation, then then executes it. [JAX](https://jax.readthedocs.io/en/latest/) is a project which combines autograd (automatic differentiation for python) and XLA to make it easier to write numpy code which runs on GPUs and lets you take derivatives automatically. This seemed like a better starting point for writing peephole improvements than trying to start from scratch because it already exists and has other useful features.
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:31]: python -> Python

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Jaxpr is JAX's intermediate representation for numpy code. While obviously different from BRIL, I found it to be pretty similar and pretty simple, so I was happy to build on top of Jaxpr rather than trying to target some other intermediate representation such as XLA or LLVM. Jaxpr, similar to BRIL, organizes programs into closures with a list of lines of code made of (basically) `(operation, in_variables, out_variables, type/shape)` tuples.
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:33]: BRIL -> Bril

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+## Hardest parts
+There were a few snags that I hit, largely through my own ignorance, before I more systematically started working through JAX's very helpful documentation, especially [Autodidax](https://jax.readthedocs.io/en/latest/autodidax.html), [How JAX primitives work](https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html), and [Writing custom interpreters in Jax](https://jax.readthedocs.io/en/latest/notebooks/Writing_custom_interpreters_in_Jax.html).
+
+### JIT
+The trickiest part of this project was when I was trying to understand the intermediate representation by tracing through and trying to directly understand the just-in-time compilation code. This was helpful for understanding some basic things about JAX, but made me spend a lot of time on the project without knowing about the existence of the much more directly helpful `make_jaxpr` and `eval_jaxpr` functions.
+
+The most useful thing that I learned from trying to read the JIT code was that JAX's JIT compilation is based on tracing all of the uses of jax's version of numpy, and then rerunning code based on that trace. Incidentally, that brings us to the next hard part...
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:43]: Is it JAX or jax (both are used in this sentence)?

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+### Jaxpr only knows about Numpy
+One issue is that Jaxpr only knows about the _numpy_ code that was run, not all the python code. As an interpreted language, Python is just running the code one line at a time. JAX doesn't reach into the interpreter to learn anything about the _python_ code, it only hears about things when you run its instrumented numpy code. This is enough to build automatic differentiation, target GPU computation, etc., but for example JAX does not learn about `if` statements.
+
+To be more explicit, when you JIT compile code using JAX it waits until the first time the function gets used, then it traces everything that function does, then it builds a Jaxpr representation based on this code. But if for example the code had an if statement that changed which code gets executed, then  JIT compiled JAX code would still run whatever version of the trace it saw first. JAX can handle non-basic blocks, but only by using things like `jax.lax.cond`.
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:48]: Add backticks around `if`?

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+# Evaluation
+My evaluation for the project was pretty straightforward, since I was focusing more on claiming that the peephole improvement was correct, rather than showing that it's useful.
+
+## Correctness
+I wrote a bunch of [tests](https://github.com/atucker/jax-peephole/tree/main/tests) which check a few basic things.
+
+1. The pattern matching for logsumexp fires when it can (even if there's code before, after, or during it), but not when the intermediate variables get used, or when the code is completely unrelated.
+2. The logsumexp peephole improvement changes the code and produces the same answer in the most basic case, or if it's happening before or after other operations, and it doesn't break anything when the intermediate variables get used (even if it doesn't actually change the code in that instance).
+3. The logsumexp peephole improvement [actually fixes](https://github.com/atucker/jax-peephole/blob/main/tests/peephole_fixes_logsumexp_overflow.out) an overflow issue! 🎉🎉🎉
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:60]: Cool!

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+Jaxpr is JAX's intermediate representation for numpy code. While obviously different from BRIL, I found it to be pretty similar and pretty simple, so I was happy to build on top of Jaxpr rather than trying to target some other intermediate representation such as XLA or LLVM. Jaxpr, similar to BRIL, organizes programs into closures with a list of lines of code made of (basically) `(operation, in_variables, out_variables, type/shape)` tuples.
+
+In order to use JAX, you need to import JAX's instrumented (among other things) version of `numpy` instead of using normal `numpy`. I've generally found this to be as easy as calling `import jax.numpy as np` and then forgetting about it.
+
+## Hardest parts
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:37]: Before jumping into the hardest parts, what do you think about an "Implementation" section to tell us more about exactly what you did? Presumably it's some kind of a pass? How does it recognize the pattern it's looking for? How flexible were you able to make the pattern-matching? That sort of thing.

File: content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+# Follow-up thoughts
+While I was working on the project, I had a few other ideas.
+
+## Where is this NaN coming from?!
+Another thing that might be useful is just making it easier to understand where the NaN is even happening. Normally once I get a NaN in my code I spend a while trying to track down exactly where it started in order to understand what the issue is, but even tracking it down can take time. Wouldn't it be nice if you could add a decorator which when added just checks every intermediate step to see if that's where the NaN happened, and then tells you if it was?
+
+While it might be tricky to do this as a peephole improvement because JAX doesn't have side-effectful code like printing, it might be a reasonable amount of effor to simply write a Jaxpr interpreter which performs this check automatically. Since you don't have to use this interpreter the rest of the time, this might be an okay implementation.
```
Comment (sampsyo) [inline content/blog/2022-05-19-numpy-peephole-improvements-with-jax.md:72]: Cool idea! Reminds me a lot of [Herbgrind](http://herbgrind.ucsd.edu):

> Herbgrind is a dynamic, binary, program analysis to find the root cause of floating point error in large programs.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Nice! This project seems very cool and perhaps very useful! Have you considered trying to get the word out—maybe it's just a matter of tweeting this blog post once it's public or something?

One suggestion you'll see inline is that I found myself wishing for more details about what the tool actually consists of. You gave us a good high-level framing, but it jumps pretty quickly to the evaluation without describing what the implementation looked like. Just a bit of detail there could help a lot.

**Comment by sampsyo** [comment]
Hi! I know it's been a while, but would you be interested in wrapping this up so we can publish it?

**Comment by atucker** [comment]
Hey, thanks for reaching out! Sorry I didn't get to this earlier -- I was in Germany then doing my A exam, but I'd be happy to wrap this up.

**Comment by sampsyo** [comment]
Hello! I am closing old issues from last year's offering to prepare for the next iteration of the course. If you would like to have this blog post published, please just leave a note and we can reopen the thread.

**Pull request closed without merging**